driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java

```diff
@@ -240,23 +245,27 @@ static MessageSettings getMessageSettings(final ConnectionDescription connection
     private static final List<String> RECOVERING_MESSAGES = asList("not master or secondary", "node is recovering");
 
     @Nullable
-    public static MongoException createSpecialException(@Nullable final BsonDocument response, final ServerAddress serverAddress,
-                                                        final String errorMessageFieldName) {
+    public static MongoException createSpecialException(@Nullable final BsonDocument response,
+                                                        final ServerAddress serverAddress,
+                                                        final String errorMessageFieldName,
+                                                        final TimeoutContext timeoutContext) {
         if (response == null) {
             return null;
         }
         int errorCode = getErrorCode(response);
         String errorMessage = getErrorMessage(response, errorMessageFieldName);
         if (ErrorCategory.fromErrorCode(errorCode) == ErrorCategory.EXECUTION_TIMEOUT) {
-            // TODO (CSOT) JAVA-5248 when timeoutMS is set, MongoOperationTimeoutException should be thrown.
+            if (timeoutContext.hasTimeoutMS()) {
+                return TimeoutContext.createMongoTimeoutException(errorMessage);
```

Comment:
We lose information here - error code / message. Is that useful information if so should the `MongoExecutionTimeoutException` be wrapped?

Reply:
Indeed, retaining an error code enhances clarity, similar to how other exceptions include one. Accordingly, I've wrapped `MongoExecutionTimeoutException` exception with a `MongoOperationTimeoutException`.

Even though the spec mentions to "transform" old exception to new one. I believe we could still use it as a cause.